src/Arcus.WebApi.Security/Authentication/CertificateAuthenticationFilter.cs

```diff
@@ -0,0 +1,171 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+using Arcus.Security.Secrets.Core.Interfaces;
+using GuardNet;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Arcus.WebApi.Security.Authentication
+{
+    /// <summary>
+    /// Authentication filter to secure HTTP requests by allowing only certain values in the client certificate.
+    /// </summary>
+    public class CertificateAuthenticationFilter : IAsyncAuthorizationFilter
+    {
+        private readonly (X509ValidationRequirement requirement, string configurationKey)[] _requirements;
```

Comment:
Why are we not creating an object for this?

Reply:
Hmm, because it's short-lived, could have not so much added value as it comes to functionality, just data that only makes sense inside the certificate validation.

Reply:
But I'm open to suggestions if you could provide an alternative.

Reply:
Actually, this is just a named Tuple.
It's in fact the same as 
```csharp
private readonly  Tuple<X509ValidationRequirement, string>[] _requirements;
```
The advantage with the code in the PR however, is that the properties of the tuple have a custom name (`requirement` and `configurationKey`) whereas when using the `Tuple<T1,T2>` class, this would be `Item1` and `Item2`.
Therefore, I think that using the named Tuple feature of C# is a better approach then using the `Tuple<T1,T2>` class and I also think that it would be overkill to create a custom type just for this.

One suggestion though:  instead of naming the properties `requirement` and `configurationKey`, I'd call them `Requirement` and `ConfigurationKey`.

Reply:
All true, except that named tuples are actually `ValueTyle<_, _>` instead of `Tuple<_, _>` ;)

Reply:
We could alias the value-tuple, [once it is supported](https://github.com/dotnet/csharplang/issues/423) :)

Reply:
That would be cool. Yet another feature that's already supported in F# ;)

Reply:
It does not bring value in terms of functionality but it's all about readability in my opinion. If you have a look at the consumer side it's not that pretty imo

Reply:
It is true that the ValueType approach is not that very obvious.   I tend to use it in private parts of classes; since it is here also exposed by the public ctor, it might indeed be a better option to create a dedicated type (which is not that of a big deal):

```csharp
public class CertificateRequirement
{
      public X509ValidationRequirement RequirementType { get; }
      public string RequirementValue { get; }
}
```

```csharp
public class CertificateAuthenticationFilter
{
     public CertificateAuthenticationFilter(IEnumerable<CertificateRequirement> requirements )
     {}
}
```

```csharp
new CertificateAuthenticationFilter(new [] 
                                    {
                                        new CertificateRequirement(X509ValidationRequirement.Subject, ""),
                                        new CertificateRequirement(X509ValidationRequirement.Thumbprint,  "")
                                    });
```

Reply:
Discussed to use named tuples only for internals which is fine, resolving